### PR TITLE
Add `hasLocationInfo` for `Type`s

### PR DESCRIPTION
**Pull request by owen-mc**
It returns a dummy location except for named types with a type declaration in the source.

### CODE REVIEW

**Review by mbg — APPROVED** [review]

**Comment by mbg** [comment]
The changes look fine, but it looks like they have made some of the tests fail and they might need updating as well.

File: go/ql/lib/semmle/go/Types.qll [hunk 1]
```diff
@@ -997,6 +1014,17 @@ class NamedType extends @namedtype, CompositeType {
...
   override Type getUnderlyingType() { result = this.getBaseType().getUnderlyingType() }
+
+  override predicate hasLocationInfo(
+    string filepath, int startline, int startcolumn, int endline, int endcolumn
+  ) {
+    exists(DeclaredType dt |
+      dt.getType() = this and
+      // Note that if the type declaration isn't in the source that we have
+      // then we use a dummy location.
+      dt.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
```
Comment (smowton) [inline go/ql/lib/semmle/go/Types.qll:1025]: I think we need to explicitly return the dummy location when there isn't such a DeclaredType

**Comment by owen-mc** [comment]
The original version was giving multiple locations for some types because it was giving the locations of type aliases as well. The refactoring seems to fix that, which is good as we don't want to get duplicate results.

**Comment by smowton** [comment]
DCA it; otherwise lgtm

**Comment by owen-mc** [comment]
@smowton The first DCA failed for unknown reasons, but the second seems absolutely uneventful, as I'd expect.

**Review by smowton — APPROVED** [review]

**Pull request merged**